stdlib/public/Cxx/CxxSequence.swift

```diff
@@ -105,8 +105,42 @@ public class CxxIterator<T>: IteratorProtocol where T: CxxSequence {
   }
 }
 
+public struct CxxIndex<T: UnsafeCxxInputIterator>: Comparable where T: Comparable {
+  fileprivate let index: T
+
+  init(_ i: T) {
+    self.index = i
+  }
+
+  public static func <(lhs: CxxIndex, rhs: CxxIndex) -> Bool { lhs.index < rhs.index }
+  public static func ==(lhs: CxxIndex, rhs: CxxIndex) -> Bool { lhs.index == rhs.index }
+}
+
 extension CxxSequence {
   public func makeIterator() -> CxxIterator<Self> {
     return CxxIterator(sequence: self)
   }
 }
+
+public protocol CxxCollection: CxxSequence, Collection where RawIterator: Comparable {
+  override associatedtype Index = CxxIndex<RawIterator>
+}
+
+extension CxxCollection {
+  public var startIndex: CxxIndex<RawIterator> {
+    var s = self
```

Comment:
This copy should _always_ get removed by the optimizer. That being said, it shouldn't be necessary in many cases. We can think about some ways to get around it.

Reply:
To be clear: if this copy isn't removed, the iterator will dangle. So we may want a stronger guarantee here.

Reply:
I guess we have two options here: require that begin has a const overload or add a shim that const-casts self.